File: native/ios/Sources/InnerVoiceWhisper/WhisperModule.swift
import Foundation
import React
import InnerVoiceWhisper

@objc(IVWhisperModule)
class IVWhisperModule: RCTEventEmitter {
    private var handle: UnsafeMutableRawPointer?
    private var currentModel: ModelInfo?

    struct ModelInfo {
        let name: String
        let url: String
        let sizeBytes: UInt64
    }

    override static func requiresMainQueueSetup() -> Bool {
        false
    }

    override func supportedEvents() -> [String]! {
        ["onTranscription"]
    }

    @objc(loadModel:resolver:rejecter:)
    func loadModel(model: NSDictionary, resolver: RCTPromiseResolveBlock, rejecter: RCTPromiseRejectBlock) {
        guard
            let name = model["name"] as? String,
            let url = model["url"] as? String,
            let size = model["sizeBytes"] as? NSNumber
        else {
            rejecter("invalid_model", "Model payload missing required fields", nil)
            return
        }

        if handle == nil {
            handle = innervoice_whisper_create(nil)
        }

        let info = innervoice_model_info_t(name: name, url: url, size_bytes: size.uint64Value)
        if innervoice_whisper_load_model(handle, info) {
            currentModel = ModelInfo(name: name, url: url, sizeBytes: size.uint64Value)
            resolver(nil)
        } else {
            let errorPtr = innervoice_whisper_last_error()
            let message = errorPtr != nil ? String(cString: errorPtr!) : "Unknown error"
            rejecter("load_failed", message, nil)
        }
    }

    @objc(transcribe:sampleRate:translate:diarize:maxThreads:resolver:rejecter:)
    func transcribe(pcmData: NSData,
                    sampleRate: NSNumber,
                    translate: NSNumber,
                    diarize: NSNumber,
                    maxThreads: NSNumber,
                    resolver: RCTPromiseResolveBlock,
                    rejecter: RCTPromiseRejectBlock) {
        guard let handle = handle else {
            rejecter("no_model", "No Whisper model loaded", nil)
            return
        }

        let frames = pcmData.length / MemoryLayout<Int16>.size
        guard frames * MemoryLayout<Int16>.size == pcmData.length else {
            rejecter("invalid_audio", "PCM payload must contain 16-bit samples", nil)
            return
        }
        let result = pcmData.bytes.bindMemory(to: Int16.self, capacity: frames)
        let transcription = innervoice_whisper_transcribe(handle,
                                                          result,
                                                          frames,
                                                          Int32(sampleRate.intValue),
                                                          translate.boolValue,
                                                          diarize.boolValue,
                                                          Int32(maxThreads.intValue))

        if let textPtr = transcription.text {
            let text = String(cString: textPtr)
            innervoice_whisper_free_result(transcription)
            sendEvent(withName: "onTranscription", body: [
                "text": text,
                "durationMs": transcription.duration_ms,
                "model": currentModel?.name ?? ""
            ])
            resolver(text)
        } else {
            let errorPtr = innervoice_whisper_last_error()
            let message = errorPtr != nil ? String(cString: errorPtr!) : "Unknown error"
            rejecter("transcription_failed", message, nil)
        }
    }

    deinit {
        if let handle = handle {
            innervoice_whisper_destroy(handle)
        }
    }
}

File: native/android/src/main/kotlin/com/innervoice/whisper/WhisperModule.kt
package com.innervoice.whisper

import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.modules.core.DeviceEventManagerModule.RCTDeviceEventEmitter

class WhisperModule(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {
    private var nativeHandle: Long = 0
    private var currentModel: Map<String, Any>? = null

    init {
        nativeHandle = nativeCreate(reactContext.cacheDir.absolutePath)
    }

    override fun getName(): String = "IVWhisperModule"

    @ReactMethod
    fun loadModel(model: Map<String, Any>, promise: Promise) {
        val name = model["name"] as? String
        val url = model["url"] as? String
        val size = (model["sizeBytes"] as? Number)?.toLong()
        if (name == null || url == null || size == null) {
            promise.reject("invalid_model", "Model payload missing required fields")
            return
        }
        if (nativeLoadModel(nativeHandle, name, url, size)) {
            currentModel = model
            promise.resolve(null)
        } else {
            promise.reject("load_failed", nativeLastError() ?: "Unknown error")
        }
    }

    @ReactMethod
    fun transcribe(pcmData: ByteArray,
                   sampleRate: Int,
                   translate: Boolean,
                   diarize: Boolean,
                   maxThreads: Int,
                   promise: Promise) {
        if (nativeHandle == 0L) {
            promise.reject("no_model", "No Whisper model loaded")
            return
        }
        if (pcmData.size % 2 != 0) {
            promise.reject("invalid_audio", "PCM payload must contain 16-bit samples")
            return
        }
        val shortArray = ShortArray(pcmData.size / 2)
        for (i in shortArray.indices) {
            val lo = pcmData[i * 2].toInt() and 0xFF
            val hi = pcmData[i * 2 + 1].toInt() and 0xFF
            shortArray[i] = (((hi shl 8) or lo)).toShort()
        }
        val result = nativeTranscribe(nativeHandle, shortArray, sampleRate, translate, diarize, maxThreads)
        if (result != null) {
            val map = Arguments.createMap()
            map.putString("text", result["text"] as? String)
            map.putDouble("durationMs", result["durationMs"] as? Double ?: 0.0)
            currentModel?.get("name")?.let { map.putString("model", it as String) }
            reactApplicationContext
                .getJSModule(RCTDeviceEventEmitter::class.java)
                .emit("onTranscription", map)
            promise.resolve(result["text"] as? String)
        } else {
            promise.reject("transcription_failed", nativeLastError() ?: "Unknown error")
        }
    }

    override fun invalidate() {
        super.invalidate()
        if (nativeHandle != 0L) {
            nativeDestroy(nativeHandle)
            nativeHandle = 0
        }
    }

    private external fun nativeCreate(cacheDir: String): Long
    private external fun nativeDestroy(handle: Long)
    private external fun nativeLoadModel(handle: Long, name: String, url: String, sizeBytes: Long): Boolean
    private external fun nativeTranscribe(handle: Long,
                                          pcm: ShortArray,
                                          sampleRate: Int,
                                          translate: Boolean,
                                          diarize: Boolean,
                                          maxThreads: Int): Map<String, Any>?
    private external fun nativeLastError(): String?

    companion object {
        init {
            System.loadLibrary("whisper_jni")
        }
    }
}
